ando/tools/generator/utils.py

```diff
@@ -67,30 +70,69 @@ def merge_dict(original_data, new_data):
         else:
-            # if the values have a simple data type and are identical then nothing needs to be done
-            if not hasattr(new_data[key], '__iter__') and original_data[key] == new_data[key]:
-                pass
+            # if the values are simple and identical then no action required
+            if not hasattr(original_data[key], '__iter__') \
+                    and new_data[key] == original_data[key]:
+                continue
             # contradicting values can not be merged
-            elif not hasattr(new_data[key], '__iter__') and original_data[key] != new_data[key]:
-                raise ValueError(f"Error different values for the same key {key} : {original_data[key]} {new_data[key]}")
+            elif not hasattr(original_data[key], '__iter__') \
+                    and new_data[key] != original_data[key]:
+                raise ValueError(f"Error different values for the same key "
+                                 f"{key}: {new_data[key]} {original_data[key]}")
             # merge lists by concatenation of values
-            if type(new_data[key]) == list:
-                result[key].extend(original_data[key])
+            if type(original_data[key]) == list:
+                result[key].extend(new_data[key])
             # merge dictionaries recursively
-            elif type(new_data[key]) == dict:
-                merge_dict(original_data[key], result[key])
+            elif type(original_data[key]) == dict:
+                result[key] = merge_dict(result[key], new_data[key])
             else:
-                raise ValueError("Can not merge unexpected data type: {type(data_existing)}")
+                raise ValueError(f"Can not merge unexpected data type: "
+                                 f"{type(original_data[key])}")
 
     return result
+
+
+def merge_dfs_by_index(original_df, new_df):
+    """
+    Merge two pandas dataframe index-by-index.
+
+    The dataframes have to share the same index name. Shared indexes will be
+    merged without data loss. In case of conflicting entries a ValueError is 
+    raised.
+
+    Parameters
+    ----------
+    original_df: dataframe
+        Pandas dataframe to be extended in-place
+    new_df: dataframe
+        Pandas dataframe with data to be added
+
+    Raises
+    ----------
+    ValueError
+        in case of incompatible index names or values
+    """
+    if original_df.index.name != new_df.index.name:
+        raise ValueError('Dataframes have incompatible indexes: '
+                         f'{original_df.index.name} != {new_df.index.name}.')
+
+    # check for contradicting values by comparing A+B with B+A
+    left_combine = original_df.combine_first(new_df)
+    right_combine = new_df.combine_first(original_df)
+
+    if not left_combine.equals(right_combine):
+        raise ValueError('Dataframes have incompatible values: '
+                         f'{left_combine.compare(right_combine)}')
+
+    return right_combine
```

Comment:
so we re not using the one already provide in pandas ?

Reply:
yes, we do as basically this is only thin wrapper for the df.combine_first() method